src/drivers/batt_smbus/batt_smbus_main.cpp

```diff
@@ -0,0 +1,485 @@
+		delete bus.dev;
+		bus.dev = nullptr;
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Start the driver.
+ *
+ * This function call only returns once the driver
+ * is either successfully up and running or failed to start.
+ */
+int start(enum BATT_SMBUS_BUS busid)
+{
+	for (unsigned i = 0; i < NUM_BUS_OPTIONS; i++) {
+		if (busid == BATT_SMBUS_BUS_ALL && bus_options[i].dev != nullptr) {
+			// This device is already started.
+			PX4_INFO("Smart battery %d already started", bus_options[i].dev);
+			continue;
+		}
+
+		if (busid != BATT_SMBUS_BUS_ALL && bus_options[i].busid != busid) {
+			// Not the one that is asked for.
+			continue;
+		}
+
+		if (start_bus(bus_options[i])) {
+			return PX4_OK;
+		}
+
+		PX4_INFO("Smart battery failed to start");
+		return PX4_ERROR;
+	}
+
+	return PX4_ERROR;
+}
+
+/**
+ * Find a bus structure for a busid.
+ */
+struct batt_smbus_bus_option &find_bus(enum BATT_SMBUS_BUS busid)
+{
+	for (unsigned i = 0; i < NUM_BUS_OPTIONS; i++) {
+		if ((busid == BATT_SMBUS_BUS_ALL || busid == bus_options[i].busid) &&
+		    bus_options[i].dev != nullptr) {
+			return bus_options[i];
+		}
+	}
+
+	errx(1, "Could not find a smart battery: Did you start it?");
+	// to satisfy other compilers
+	return bus_options[0];
+}
+
+
+
+int manufacture_date()
+{
+	int result = -1;
+	uint16_t man_date = 0;
+
+	for (uint8_t i = 0; i < NUM_BUS_OPTIONS; i++) {
+		struct batt_smbus_bus_option &bus = bus_options[i];
+
+		if (bus.dev != nullptr) {
+			PX4_INFO("%s", bus.devpath);
+			result = bus.dev->manufacture_date(&man_date);
+		}
+	}
+
+	if (PX4_OK == result) {
+		// Convert the uint16_t into human-readable date format.
+		uint16_t year = ((man_date >> 9) & 0xFF) + 1980;
+		uint8_t month = (man_date >> 5) & 0xF;
+		uint8_t day = man_date & 0x1F;
+		PX4_INFO("The manufacturer date is: %4d-%02d-%02d", year, month, day);
+		return PX4_OK;
+
+	} else {
+		PX4_WARN("Unable to read the manufacturer date.");
+		return PX4_ERROR;
+	}
+
+}
+
+int manufacturer_name()
+{
+	int result = -1;
+	uint8_t man_name[22];
+
+	for (uint8_t i = 0; i < NUM_BUS_OPTIONS; i++) {
+		struct batt_smbus_bus_option &bus = bus_options[i];
+
+		if (bus.dev != nullptr) {
+			PX4_INFO("%s", bus.devpath);
+			result = bus.dev->manufacturer_name(man_name, sizeof(man_name));
+		}
+	}
+
+	if (PX4_OK == result) {
+		PX4_INFO("The manufacturer name: %s", man_name);
+		return PX4_OK;
+
+	} else {
+		PX4_WARN("Unable to read manufacturer name.");
+		return PX4_ERROR;
+	}
+}
+
+int serial_number()
+{
+	uint16_t serial_num = 0;
+
+	for (uint8_t i = 0; i < NUM_BUS_OPTIONS; i++) {
+		struct batt_smbus_bus_option &bus = bus_options[i];
+
+		if (bus.dev != nullptr) {
+			PX4_INFO("%s", bus.devpath);
+			serial_num = bus.dev->get_serial_number();
+		}
+	}
+
+	PX4_INFO("The serial number: 0x%04x (%d in decimal)", serial_num, serial_num);
+
+	return PX4_OK;
+}
+
+void usage()
+{
+	PX4_INFO("\n\tstart            	 	- Starts the driver."
+		 "\n\tstop 				- Stops the driver."
```

Comment:
You might want to look at the helpers in px4_module.h for doing this.
https://github.com/PX4/Firmware/blob/master/src/platforms/px4_module.h#L404-L494

Take a look at the GPS driver as an example. https://github.com/PX4/Firmware/blob/master/src/drivers/gps/gps.cpp#L927-L967